File: backend/app/models/cross_engine.py
"""Cross-Engine Link Resolution API models.

Gap 5-3: Cross-Engine Correlation Links

Pydantic models for cross-engine link API endpoints.
"""

from pydantic import BaseModel, ConfigDict, Field


# =============================================================================
# Cross-Linked Item Models
# =============================================================================


class CrossLinkedTimelineEventModel(BaseModel):
    """Timeline event with minimal data for cross-engine linking."""

    model_config = ConfigDict(populate_by_name=True)

    event_id: str = Field(..., alias="eventId", description="Timeline event UUID")
    event_date: str = Field(..., alias="eventDate", description="Event date (ISO format)")
    event_type: str = Field(..., alias="eventType", description="Event type classification")
    description: str = Field(..., description="Event description")
    document_id: str | None = Field(
        None, alias="documentId", description="Source document UUID"
    )
    document_name: str | None = Field(
        None, alias="documentName", description="Source document filename"
    )
    source_page: int | None = Field(
        None, alias="sourcePage", description="Source page number"
    )
    confidence: float = Field(default=1.0, ge=0.0, le=1.0, description="Confidence score")


class CrossLinkedContradictionModel(BaseModel):
    """Contradiction with minimal data for cross-engine linking."""

    model_config = ConfigDict(populate_by_name=True)

    contradiction_id: str = Field(
        ..., alias="contradictionId", description="Contradiction UUID"
    )
    contradiction_type: str = Field(
        ..., alias="contradictionType", description="Type of contradiction"
    )
    severity: str = Field(..., description="Severity level: high, medium, low")
    explanation: str = Field(..., description="Explanation of the contradiction")
    statement_a_excerpt: str = Field(
        ..., alias="statementAExcerpt", description="First statement excerpt"
    )
    statement_b_excerpt: str = Field(
        ..., alias="statementBExcerpt", description="Second statement excerpt"
    )
    document_a_id: str = Field(
        ..., alias="documentAId", description="First document UUID"
    )
    document_a_name: str = Field(
        ..., alias="documentAName", description="First document filename"
    )
    document_b_id: str = Field(
        ..., alias="documentBId", description="Second document UUID"
    )
    document_b_name: str = Field(
        ..., alias="documentBName", description="Second document filename"
    )
    confidence: float = Field(default=0.5, ge=0.0, le=1.0, description="Confidence score")


class CrossLinkedEntityModel(BaseModel):
    """Entity with minimal data for cross-engine linking."""

    model_config = ConfigDict(populate_by_name=True)

    entity_id: str = Field(..., alias="entityId", description="Entity UUID")
    canonical_name: str = Field(
        ..., alias="canonicalName", description="Entity canonical name"
    )
    entity_type: str = Field(..., alias="entityType", description="Entity type")
    aliases: list[str] = Field(default_factory=list, description="Entity aliases")


# =============================================================================
# Response Models
# =============================================================================


class EntityJourneyResponse(BaseModel):
    """Response for entity journey (timeline events for an entity)."""

    model_config = ConfigDict(populate_by_name=True)

    entity_id: str = Field(..., alias="entityId", description="Entity UUID")
    entity_name: str = Field(..., alias="entityName", description="Entity canonical name")
    entity_type: str = Field(..., alias="entityType", description="Entity type")
    events: list[CrossLinkedTimelineEventModel] = Field(
        default_factory=list, description="Timeline events involving this entity"
    )
    total_events: int = Field(
        default=0, alias="totalEvents", description="Total number of events"
    )
    date_range_start: str | None = Field(
        None, alias="dateRangeStart", description="Earliest event date"
    )
    date_range_end: str | None = Field(
        None, alias="dateRangeEnd", description="Latest event date"
    )


class EntityContradictionSummaryResponse(BaseModel):
    """Response for entity contradictions summary."""

    model_config = ConfigDict(populate_by_name=True)

    entity_id: str = Field(..., alias="entityId", description="Entity UUID")
    entity_name: str = Field(..., alias="entityName", description="Entity canonical name")
    contradictions: list[CrossLinkedContradictionModel] = Field(
        default_factory=list, description="Contradictions involving this entity"
    )
    total_contradictions: int = Field(
        default=0, alias="totalContradictions", description="Total contradiction count"
    )
    high_severity_count: int = Field(
        default=0, alias="highSeverityCount", description="High severity count"
    )
    medium_severity_count: int = Field(
        default=0, alias="mediumSeverityCount", description="Medium severity count"
    )
    low_severity_count: int = Field(
        default=0, alias="lowSeverityCount", description="Low severity count"
    )


class TimelineEventContextResponse(BaseModel):
    """Response for timeline event context."""

    model_config = ConfigDict(populate_by_name=True)

    event_id: str = Field(..., alias="eventId", description="Event UUID")
    event_date: str = Field(..., alias="eventDate", description="Event date")
    event_type: str = Field(..., alias="eventType", description="Event type")
    description: str = Field(..., description="Event description")
    document_id: str | None = Field(
        None, alias="documentId", description="Source document UUID"
    )
    document_name: str | None = Field(
        None, alias="documentName", description="Source document filename"
    )
    entities: list[CrossLinkedEntityModel] = Field(
        default_factory=list, description="Entities involved in this event"
    )
    related_contradictions: list[CrossLinkedContradictionModel] = Field(
        default_factory=list,
        alias="relatedContradictions",
        description="Contradictions for entities in this event",
    )


class ContradictionContextResponse(BaseModel):
    """Response for contradiction context."""

    model_config = ConfigDict(populate_by_name=True)

    contradiction_id: str = Field(
        ..., alias="contradictionId", description="Contradiction UUID"
    )
    entity_id: str = Field(..., alias="entityId", description="Entity UUID")
    entity_name: str = Field(..., alias="entityName", description="Entity canonical name")
    contradiction_type: str = Field(
        ..., alias="contradictionType", description="Type of contradiction"
    )
    severity: str = Field(..., description="Severity level")
    explanation: str = Field(..., description="Explanation")
    related_events: list[CrossLinkedTimelineEventModel] = Field(
        default_factory=list,
        alias="relatedEvents",
        description="Timeline events for the same entity",
    )


# =============================================================================
# Error Response Model
# =============================================================================


class CrossEngineErrorDetail(BaseModel):
    """Error detail structure for cross-engine API."""

    code: str = Field(..., description="Machine-readable error code")
    message: str = Field(..., description="Human-readable error message")
    details: dict = Field(default_factory=dict, description="Additional error context")


class CrossEngineErrorResponse(BaseModel):
    """Error response structure for cross-engine API."""

    error: CrossEngineErrorDetail

File: frontend/src/types/crossEngine.ts
/**
 * Cross-Engine Link Types
 *
 * Gap 5-3: Cross-Engine Correlation Links
 *
 * Types for cross-engine data correlation enabling navigation between:
 * - Timeline events ↔ Entities
 * - Timeline events ↔ Contradictions
 * - Entities ↔ Contradictions
 */

// =============================================================================
// Cross-Linked Item Types
// =============================================================================

/**
 * Timeline event with minimal data for cross-engine linking
 */
export interface CrossLinkedTimelineEvent {
  eventId: string;
  eventDate: string;
  eventType: string;
  description: string;
  documentId: string | null;
  documentName: string | null;
  sourcePage: number | null;
  confidence: number;
}

/**
 * Contradiction with minimal data for cross-engine linking
 */
export interface CrossLinkedContradiction {
  contradictionId: string;
  contradictionType: string;
  severity: 'high' | 'medium' | 'low';
  explanation: string;
  statementAExcerpt: string;
  statementBExcerpt: string;
  documentAId: string;
  documentAName: string;
  documentBId: string;
  documentBName: string;
  confidence: number;
}

/**
 * Entity with minimal data for cross-engine linking
 */
export interface CrossLinkedEntity {
  entityId: string;
  canonicalName: string;
  entityType: string;
  aliases: string[];
}

// =============================================================================
// Response Types
// =============================================================================

/**
 * Response for entity journey (timeline events for an entity)
 */
export interface EntityJourneyResponse {
  entityId: string;
  entityName: string;
  entityType: string;
  events: CrossLinkedTimelineEvent[];
  totalEvents: number;
  dateRangeStart: string | null;
  dateRangeEnd: string | null;
}

/**
 * Response for entity contradictions summary
 */
export interface EntityContradictionSummary {
  entityId: string;
  entityName: string;
  contradictions: CrossLinkedContradiction[];
  totalContradictions: number;
  highSeverityCount: number;
  mediumSeverityCount: number;
  lowSeverityCount: number;
}

/**
 * Response for timeline event context
 */
export interface TimelineEventContext {
  eventId: string;
  eventDate: string;
  eventType: string;
  description: string;
  documentId: string | null;
  documentName: string | null;
  entities: CrossLinkedEntity[];
  relatedContradictions: CrossLinkedContradiction[];
}

/**
 * Response for contradiction context
 */
export interface ContradictionContext {
  contradictionId: string;
  entityId: string;
  entityName: string;
  contradictionType: string;
  severity: string;
  explanation: string;
  relatedEvents: CrossLinkedTimelineEvent[];
}

// =============================================================================
// Link Types for UI Components
// =============================================================================

/**
 * Types of cross-engine links that can be displayed
 */
export type CrossLinkType =
  | 'entity-timeline'      // Link to entity's timeline journey
  | 'entity-contradiction' // Link to entity's contradictions
  | 'event-entities'       // Link to entities in an event
  | 'event-contradiction'  // Link to contradictions related to event
  | 'contradiction-events' // Link to events related to a contradiction
  | 'contradiction-entity';// Link to entity involved in contradiction

/**
 * Cross-link metadata for rendering
 */
export interface CrossLink {
  type: CrossLinkType;
  targetId: string;
  targetTab: 'timeline' | 'entities' | 'contradictions';
  label: string;
  count?: number;
}

/**
 * Cross-link navigation parameters
 */
export interface CrossLinkNavigation {
  matterId: string;
  targetTab: 'timeline' | 'entities' | 'contradictions';
  entityId?: string;
  eventId?: string;
  contradictionId?: string;
  highlight?: string;
}

// =============================================================================
// Counts for badges
// =============================================================================

/**
 * Cross-engine counts for displaying badges on items
 */
export interface CrossEngineCounts {
  /** Number of timeline events (for entities) */
  timelineEventCount?: number;
  /** Number of contradictions (for entities or events) */
  contradictionCount?: number;
  /** Number of high-severity contradictions */
  highSeverityCount?: number;
  /** Number of related entities (for events or contradictions) */
  entityCount?: number;
  /** Date range (for entity journey) */
  dateRange?: {
    start: string | null;
    end: string | null;
  };
}

// =============================================================================
// Story 5.4: Cross-Engine Consistency Issue Types
// =============================================================================

/**
 * Issue type classification
 */
export type ConsistencyIssueType =
  | 'date_mismatch'
  | 'entity_name_mismatch'
  | 'amount_discrepancy'
  | 'citation_conflict'
  | 'timeline_gap'
  | 'duplicate_event';

/**
 * Issue severity levels
 */
export type ConsistencyIssueSeverity = 'info' | 'warning' | 'error';

/**
 * Issue resolution status
 */
export type ConsistencyIssueStatus = 'open' | 'reviewed' | 'resolved' | 'dismissed';

/**
 * Engine type
 */
export type EngineType = 'timeline' | 'entity' | 'citation' | 'contradiction' | 'rag';

/**
 * A cross-engine consistency issue
 */
export interface ConsistencyIssue {
  id: string;
  matterId: string;
  issueType: ConsistencyIssueType;
  severity: ConsistencyIssueSeverity;
  sourceEngine: EngineType;
  sourceId: string | null;
  sourceValue: string | null;
  conflictingEngine: EngineType;
  conflictingId: string | null;
  conflictingValue: string | null;
  description: string;
  documentId: string | null;
  documentName: string | null;
  status: ConsistencyIssueStatus;
  resolvedBy: string | null;
  resolvedAt: string | null;
  resolutionNotes: string | null;
  detectedAt: string;
  createdAt: string;
  updatedAt: string;
  metadata: Record<string, unknown>;
}

/**
 * Consistency issue summary counts
 */
export interface ConsistencyIssueSummary {
  totalCount: number;
  openCount: number;
  warningCount: number;
  errorCount: number;
}

/**
 * List response for consistency issues
 */
export interface ConsistencyIssueListResponse {
  data: ConsistencyIssue[];
  meta: {
    limit: number;
    offset: number;
    count: number;
  };
}

/**
 * Summary response for consistency issues
 */
export interface ConsistencyIssueSummaryResponse {
  data: ConsistencyIssueSummary;
}

/**
 * Result of running a consistency check
 */
export interface ConsistencyCheckResult {
  issuesFound: number;
  issuesCreated: number;
  enginesChecked: string[];
  durationMs: number;
}

/**
 * Labels for issue types
 */
export const ISSUE_TYPE_LABELS: Record<ConsistencyIssueType, string> = {
  date_mismatch: 'Date Mismatch',
  entity_name_mismatch: 'Entity Name Mismatch',
  amount_discrepancy: 'Amount Discrepancy',
  citation_conflict: 'Citation Conflict',
  timeline_gap: 'Timeline Gap',
  duplicate_event: 'Duplicate Event',
};

/**
 * Descriptions for issue types
 */
export const ISSUE_TYPE_DESCRIPTIONS: Record<ConsistencyIssueType, string> = {
  date_mismatch: 'Different dates found for the same event in timeline and entity data',
  entity_name_mismatch: 'Entity name variations between MIG and citations',
  amount_discrepancy: 'Monetary or numeric values differ between extractions',
  citation_conflict: 'Citation information conflicts with document content',
  timeline_gap: 'Missing events in timeline that are referenced elsewhere',
  duplicate_event: 'Same event appears multiple times with different details',
};

/**
 * Labels for severity levels
 */
export const SEVERITY_LABELS: Record<ConsistencyIssueSeverity, string> = {
  info: 'Info',
  warning: 'Warning',
  error: 'Error',
};

/**
 * Labels for issue status
 */
export const STATUS_LABELS: Record<ConsistencyIssueStatus, string> = {
  open: 'Open',
  reviewed: 'Reviewed',
  resolved: 'Resolved',
  dismissed: 'Dismissed',
};
